File: problems/Binary Tree from inorder and postorder.js
// leetcode:  https://leetcode.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal/discuss/491447/JavaScript-Solution

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} inorder
 * @param {number[]} postorder
 * @return {TreeNode}
 */
function TreeNode(val) {
  this.val = val;
  this.left = this.right = null;
  return this;
}

var buildTree = function(inOrder, postOrder) {
  function helper(start, end) {
    if (start > end) return null ;

    const val = postOrder.pop(); // remove the root
    console.log(val, start, end);

    const tree = new TreeNode(val);
    // build right tree first as thats how the Post Order Array is built, if read backward
    // get index of root val from the inOrder array, as that splits the array in LHS tree and RHS tree
    tree.right = helper(inOrder.indexOf(val) + 1, end);
    tree.left = helper(start, inOrder.indexOf(val) - 1);

    return tree;
  }

  return helper(0, inOrder.length - 1);
};

const inorder = [9, 3, 15, 20, 7];
const postorder = [9, 15, 7, 20, 3];
/**
 *  return
 *     3
      /   \
     9     20
          /  \
        15    7
 */

const ans = buildTree(inorder, postorder);
console.log(ans);

File: problems/Binary Tree from inorder and postorder.py

# https://leetcode.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal
# https://leetcode.com/explore/learn/card/data-structure-tree/133/conclusion/942/discuss/153572/Python-Recursive-and-Iterative

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def buildTree(self, inOrder, postOrder):

        def helper(start, end):
            if start > end:
                return None

            if len(postOrder) > 0:
                val = postOrder.pop()
                tree = TreeNode(val)
                valIdx = inOrder.index(val)
                print(val, start, end)

                # build the tree, right branch first, that is how postorder trees are built
                tree.right = helper(valIdx+1, end)
                tree.left = helper(start, valIdx-1)

                return tree

        return helper(0, len(inOrder)-1)


inorder = [9, 3, 15, 20, 7]
postorder = [9, 15, 7, 20, 3]
# /**
#  *  return
#  *     3
#       /   \
#      9     20
#           /  \
#         15    7
#  */

ans = Solution().buildTree(inorder, postorder)
print(ans)
